source/iterator/worker.go

```diff
@@ -0,0 +1,350 @@
+func (w *Worker) HasNext(ctx context.Context) (bool, error) {
+	if w.rows != nil && w.rows.Next() {
+		return true, nil
+	}
+
+	if err := w.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	if w.rows.Next() {
+		return true, nil
+	}
+
+	// At this point, there are no more rows to load
+	// so if we are in snapshot mode, we can switch to CDC
+	if w.position.LatestSnapshotValue != nil {
+		// switch to CDC mode
+		w.position.LastProcessedValue = w.position.LatestSnapshotValue
+		w.position.LatestSnapshotValue = nil
+
+		w.iterator.updateTablePosition(w.table, w.position)
+
+		// and load new rows
+		if err := w.loadRows(ctx); err != nil {
+			return false, fmt.Errorf("load rows: %w", err)
+		}
+
+		return w.rows.Next(), nil
+	}
+
+	return false, nil
+}
+
+// Next returns the next record.
+func (w *Worker) Next(_ context.Context) (opencdc.Record, error) {
+	row := make(map[string]any)
+	if err := w.rows.MapScan(row); err != nil {
+		return opencdc.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := columntypes.TransformRow(row, w.columnTypes)
+	if err != nil {
+		return opencdc.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	if _, ok := transformedRow[w.orderingColumn]; !ok {
+		return opencdc.Record{}, fmt.Errorf("ordering column %q not found", w.orderingColumn)
+	}
+
+	key := make(opencdc.StructuredData)
+	for i := range w.keyColumns {
+		val, ok := transformedRow[w.keyColumns[i]]
+		if !ok {
+			return opencdc.Record{}, fmt.Errorf("key column %q not found", w.keyColumns[i])
+		}
+
+		key[w.keyColumns[i]] = val
+	}
+
+	rowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return opencdc.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	// set a new position into the variable,
+	// to avoid saving position into the struct until we marshal the position
+	position := w.iterator.getPosition()
+
+	tablePos := w.getTablePosition(position, transformedRow)
+
+	// set the value from iter.orderingColumn column you chose
+	position.TablePositions[w.table] = tablePos
+
+	sdkPosition, err := position.marshal()
+	if err != nil {
+		return opencdc.Record{}, fmt.Errorf("failed converting to SDK position :%w", err)
+	}
+
+	w.position = position.TablePositions[w.table]
+	w.iterator.updateTablePosition(w.table, tablePos)
+
+	metadata := opencdc.Metadata{
+		opencdc.MetadataCollection: w.table,
+	}
+	metadata.SetCreatedAt(time.Now().UTC())
+
+	if w.position.LatestSnapshotValue != nil {
+		return sdk.Util.Source.NewRecordSnapshot(sdkPosition, metadata, key, opencdc.RawData(rowBytes)), nil
+	}
+
+	return sdk.Util.Source.NewRecordCreate(sdkPosition, metadata, key, opencdc.RawData(rowBytes)), nil
+}
+
+// Stop stops worker.
+func (w *Worker) Stop() error {
+	var err error
+
+	if w.rows != nil {
+		err = w.rows.Close()
+	}
+
+	if err != nil {
+		return fmt.Errorf("close db rows: %w", err)
+	}
+
```

Comment:
Nitpick: this can in the body of the `if` statement, i.e.:

```go
if w.rows != nil {
    err := w.Rows.Close()
    if err != nil {
        return fmt.Errorf()
    }
}
```

It's the same result, but the scope of `err` is smaller.